### PR TITLE
Polling Improvements

### PR DESCRIPTION
**Pull request by xTEddie**
# PR Details

### Description

_Include a description of the problem to be solved_

## Solution Proposed

_Detail what is the solution proposed, include links to design document if required or any other document required to support the solution_

### Acceptance criteria

_Define what are the conditions to consider the PR has achieved the intended goal_

## Test cases and evidence

_Include what tests cases were considered, any evidence of testing for future references, to identify any corner cases, etc_

### Sanity Tests

- [ ] You have tested all changes in Popout mode
- [ ] You have tested all changes in cross browsers i.e Edge, Chrome, Firefox, Safari and mobile devices(iOS and Android)
- [ ] Your changes are included in the [CHANGELOG](../CHANGE_LOG.md)

## A11y

- [ ] You have run the [Automated Accessibility Check](https://accessibilityinsights.io/docs/en/windows/getstarted/automatedchecks) and have no reported issues

***Please provide justification if any of the validations has been skipped.***

### CODE REVIEW

File: src/core/messaging/ACSClient.ts [hunk 1]
```diff
@@ -173,50 +185,53 @@ export class ACSConversation {
...
-                }, delay.done === true ? 10000 : delay.value);
+                }, delay.done === true ? defaultInterval : delay.value);
             };
 
-            // Poll messages until WS established connection
+            this.keepPolling = true;
             const delayGenerator = nextDelay();
-            await pollForMessages(delayGenerator);
+            if (optionalParams.disablePolling === false) {
+                await pollForMessages(delayGenerator);
```
Comment (mithunjohny) [inline src/core/messaging/ACSClient.ts:232]: should we add const delayGenerator = nextDelay(); within if condition

File: src/commands/startPolling.ts [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @param conversation ACS Conversation object
+ */
+const startPolling = async (isInitialized: boolean, scenarioMarker: ScenarioMarker, liveChatVersion: number, requestId: string, chatId: string, conversation: ACSConversation | null): Promise<void> => {
+    scenarioMarker.startScenario(TelemetryEvent.StartPolling);
+
+    if (!isInitialized) {
+        exceptionThrowers.throwUninitializedChatSDK(scenarioMarker, TelemetryEvent.StartPolling);
+    }
+
+    if (liveChatVersion === LiveChatVersion.V2) {
```
Comment (elopezanaya) [inline src/commands/startPolling.ts:24]: do we still need v2 validation?, if we do, then lets  complete the scenario marker when not v2

File: src/core/messaging/ACSClient.ts (hunk 1 above)
Comment (xTEddie) [reply src/core/messaging/ACSClient.ts:232]: Fixed

File: src/commands/stopPolling.ts [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+ * @param conversation ACS Conversation object
+ */
+const stopPolling = async (isInitialized: boolean, scenarioMarker: ScenarioMarker, liveChatVersion: number, requestId: string, chatId: string, conversation: ACSConversation | null): Promise<void> => {
+    scenarioMarker.startScenario(TelemetryEvent.StopPolling);
+
+    if (!isInitialized) {
+        exceptionThrowers.throwUninitializedChatSDK(scenarioMarker, TelemetryEvent.StopPolling);
+    }
+
+    if (liveChatVersion === LiveChatVersion.V2) {
```
Comment (elopezanaya) [inline src/commands/stopPolling.ts:24]: same comment about the need to use v2

File: src/core/messaging/ACSClient.ts [hunk 4]
```diff
@@ -173,50 +185,53 @@ export class ACSConversation {
...
                 const delay = delayGenerator.next();
-                setTimeout(() => {
+                this.pollingTimer = setTimeout(() => {
                     pollForMessages(delayGenerator);
-                }, delay.done === true ? 10000 : delay.value);
+                }, delay.done === true ? defaultInterval : delay.value);
             };
 
-            // Poll messages until WS established connection
+            this.keepPolling = true;
```
Comment (elopezanaya) [inline src/core/messaging/ACSClient.ts:229]: this flag should be handled only for start/stop functions,

File: src/core/messaging/ACSClient.ts [hunk 5]
```diff
@@ -173,50 +185,53 @@ export class ACSConversation {
         return participants;
     }
 
-    public async registerOnNewMessage(onNewMessageCallback: CallableFunction): Promise<void> {
+    public async registerOnNewMessage(onNewMessageCallback: CallableFunction, optionalParams: ACSRegisterOnNewMessageOptionalParams = {disablePolling: false}): Promise<void> {
```
Comment (elopezanaya) [inline src/core/messaging/ACSClient.ts:188]: do we really need to pass the whole optional params when we need only 1 flag?

File: src/core/messaging/ACSClient.ts [hunk 6]
```diff
@@ -376,6 +391,11 @@ export class ACSConversation {
                 });
             }
 
+            await this.stopPolling();
+            if (this.pollingTimer) {
+                clearTimeout(this.pollingTimer as number);
```
Comment (elopezanaya) [inline src/core/messaging/ACSClient.ts:396]: 👍

**Review by elopezanaya — CHANGES_REQUESTED** [review]

File: src/core/messaging/ACSClient.ts (hunk 5 above)
Comment (mithunjohny) [reply src/core/messaging/ACSClient.ts:188]: i think it is fine since our pattern is to pass an object everywhere

File: src/commands/startPolling.ts (hunk 2 above)
Comment (xTEddie) [reply src/commands/startPolling.ts:24]: Yes, as safety measure

File: src/core/messaging/ACSClient.ts (hunk 5 above)
Comment (xTEddie) [reply src/core/messaging/ACSClient.ts:188]: Yes, it's to keep the `backward compatibility` as well. The signature has not changed since the last param is optional so it would not break existing usage of that method.

**Review by mithunjohny — APPROVED** [review]

**Review by elopezanaya — APPROVED** [review]
